source/distributed/tria.cc

```diff
@@ -1213,42 +1238,112 @@ namespace parallel
+            *data_fixed_it =
               Utilities::pack(cell_status, /*allow_compression=*/false);
-            ++data_it;
+            ++data_fixed_it;
 
             // Proceed with all registered callback functions.
             // Skip cells with the CELL_INVALID flag.
             if (cell_status !=
                 parallel::distributed::Triangulation<dim,
                                                      spacedim>::CELL_INVALID)
               {
-                for (auto callback_it = pack_callbacks.cbegin();
-                     callback_it != pack_callbacks.cend();
-                     ++callback_it, ++data_it)
+                // Pack fixed size data.
+                for (auto callback_it = pack_callbacks_fixed.cbegin();
+                     callback_it != pack_callbacks_fixed.cend();
+                     ++callback_it, ++data_fixed_it)
+                  {
+                    *data_fixed_it = (*callback_it)(dealii_cell, cell_status);
+                  }
+
+                // Pack variable size data.
+                // If we store variable size data, we need to transfer
+                // the sizes of each corresponding callback function
+                // via fixed size transfer as well.
+                if (variable_transfer)
                   {
-                    *data_it = (*callback_it)(dealii_cell, cell_status);
+                    const unsigned int n_variable_size_data_sets_on_cell =
+                      ((cell_status ==
+                        parallel::distributed::Triangulation<dim, spacedim>::
+                          CELL_INVALID) ?
+                         0 :
+                         n_callbacks_variable);
+                    data_cell_variable_it->resize(
+                      n_variable_size_data_sets_on_cell);
+
+                    auto callback_it      = pack_callbacks_variable.cbegin();
+                    auto data_variable_it = data_cell_variable_it->begin();
+                    auto sizes_variable_it =
+                      cell_sizes_variable_cumulative.begin();
+                    for (; callback_it != pack_callbacks_variable.cend();
+                         ++callback_it, ++data_variable_it, ++sizes_variable_it)
+                      {
+                        *data_variable_it =
+                          (*callback_it)(dealii_cell, cell_status);
+
+                        // Store data sizes for each callback function first.
+                        // Make it cumulative below.
+                        *sizes_variable_it = data_variable_it->size();
+                      }
+
+                    // Turn size vector into its cumulative representation.
+                    std::partial_sum(cell_sizes_variable_cumulative.begin(),
+                                     cell_sizes_variable_cumulative.end(),
+                                     cell_sizes_variable_cumulative.begin());
+
+                    // Serialize cumulative size vector value-by-value.
```

Comment:
can you say here why you do it this way? namely, because the size is fixed and you don't want to serialize the vector with its overhead?